### PR TITLE
Move multicast discovery to a plugin

### LINKED ISSUE #13019: make multicast zen discovery a plugin
Followup to #12999 which explains the rationale.

### PR DESCRIPTION
**Pull request by rjernst**
Multicast has known issues (see #12999 and #12993). This change moves
multicast into a plugin, and deprecates it in the docs.  It also allows
for plugging in multiple zen ping implementations.

I think we should also state in the docs this is deprecated, but I didn't know how to word that or how to mark something as deprecated in asciidoc.

closes #13019

### CODE REVIEW

**Comment by rmuir** [comment]
The plugin works, I installed it and set the boolean and things joined over the network.
The plugin needs to be added to the help file with the other plugins right?
Also, I think you have to add it to qa/smoke-test-plugins or it should fail.

File: plugins/cloud-gce/src/main/java/org/elasticsearch/discovery/gce/GceDiscovery.java [hunk 1]
```diff
@@ -46,7 +46,6 @@ public GceDiscovery(Settings settings, ClusterName clusterName, ThreadPool threa
         super(settings, clusterName, threadPool, transportService, clusterService, nodeSettingsService,
                 pingService, electMasterService, discoverySettings);
 
-        // TODO Add again force disable multicast
         // See related issue in AWS plugin https://github.com/elastic/elasticsearch-cloud-aws/issues/179
```
Comment (dadoonet) [inline plugins/cloud-gce/src/main/java/org/elasticsearch/discovery/gce/GceDiscovery.java]: You can also remove the second comment.

**Comment by rjernst** [comment]
I pushed new commits. mvn verify passes.

**Comment by rjernst** [comment]
I also ran verify with -Des.node.mode=network

**Comment by rmuir** [comment]
me too. +1

**Comment by dadoonet** [comment]
Missing change in https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/plugins/PluginManager.java and in plugin manager tests.

**Comment by bleskes** [comment]
LGTM2, apart from what @dadoonet said (which I wouldn't have come up with my self ... )

**Comment by dadoonet** [comment]
Yeah we need to find a way to automatically fail when we omit such a change.
May be a qa thing for each plugin listed in plugins?

**Comment by dadoonet** [comment]
@rjernst Also the documentation build is failing. Could you check?

**Comment by dadoonet** [comment]
> Also the documentation build is failing. Could you check?

Thank you @jpountz for fixing it: c54a4c4f8995fea61f0e02586faf6101661231e4

**Comment by brwe** [comment]
@rjernst I think this should this be backported to 2.0 branch?

**Comment by rjernst** [comment]
Yes, I wanted to give CI a little time to chew. I'll backport now.

**Comment by dadoonet** [comment]
@rjernst Could you also add `discovery-multicast` to [PluginManagerIT](https://github.com/elastic/elasticsearch/blob/master/core/src/test/java/org/elasticsearch/plugins/PluginManagerIT.java#L551-L574)?

**Comment by s1monw** [comment]
> @rjernst Could you also add discovery-multicast to PluginManagerIT?

done on master and 2.0

**Comment by clintongormley** [comment]
@rjernst Please could you add this to the migration docs as well

**Pull request merged**